include/swift/SIL/SILConstants.h

```diff
@@ -0,0 +1,458 @@
+/// avoid pointless memory bloat + copying.  This is intended to be a
+/// light-weight POD type we can put in hash tables and pass around by-value.
+///
+/// Internally, this value has multiple ways to represent the same sorts of
+/// symbolic values (e.g. to save memory).  It provides a simpler public
+/// interface though.
+class SymbolicValue {
+private:
+  enum RepresentationKind {
+    /// This value is an alloc stack that is has not (yet) been initialized
+    /// by flow-sensitive analysis.
+    RK_UninitMemory,
+
+    /// This symbolic value cannot be determined, carries multiple values
+    /// (i.e., varies dynamically at the top level), or is of some type that
+    /// we cannot analyze and propagate (e.g. NSObject).
+    ///
+    RK_Unknown,
+
+    /// This value is known to be a metatype reference.  The type is stored
+    /// in the "metatype" member.
+    RK_Metatype,
+
+    /// This value is known to be a function reference, e.g. through
+    /// function_ref directly, or a devirtualized method reference.
+    RK_Function,
+
+    /// This value is represented with a bump-pointer allocated APInt.
+    RK_Integer,
+
+    /// This value is represented with an inline integer representation.
+    RK_IntegerInline,
+
+    /// This value is represented with a bump-pointer allocated APFloat.
+    RK_Float,
+
+    /// This value is representation with an inline float representation.
+    RK_Float32,
+
+    /// This value is representation with an inline double representation.
+    RK_Float64,
+
+    /// This value is represented with a bump-pointer allocated char array
+    /// representing a UTF-8 encoded string.
+    RK_String,
+
+    /// This value is an array, struct, or tuple of constants.  This is
+    /// tracked by the "aggregate" member of the value union.
+    RK_Aggregate,
+
+    /// This value is an enum with no payload.
+    RK_Enum,
+
+    /// This value is an enum with a payload.
+    RK_EnumWithPayload,
+
+    /// This represents a direct reference to the address of a memory object.
+    RK_DirectAddress,
+
+    /// This represents an index *into* a memory object.
+    RK_DerivedAddress,
+
+    /// This represents an array value.
+    RK_Array,
+
+    /// This represents an address of a memory object containing an array.
+    RK_ArrayAddress,
+  };
+
+  union {
+    /// When the value is Unknown, this contains information about the
+    /// unfoldable part of the computation.
+    UnknownSymbolicValue *unknown;
+
+    /// This is always a SILType with an object category.  This is the value
+    /// of the underlying instance type, not the MetatypeType.
+    TypeBase *metatype;
+
+    SILFunction *function;
+
+    /// When this SymbolicValue is of "Inst" kind, this contains a
+    /// pointer to the instruction whose value this holds.  This is known to
+    /// be one of a closed set of constant instructions:
+    ///    IntegerLiteralInst, FloatLiteralInst, StringLiteralInst
+    SingleValueInstruction *inst;
```

Comment:
@marcrasi  Looks like this `inst` representation for literals is not used anymore. Is this dead code? (and also the declaration of SingleValueInstruction at the beginning of the file).

Reply:
You are right. I forgot to remove it from the branch code. @marcrasi please feel free to remove.